hazelcast/sql.py

```diff
@@ -0,0 +1,1593 @@
+import logging
+import uuid
+from threading import RLock
+
+from hazelcast.errors import HazelcastError
+from hazelcast.future import Future, ImmediateFuture
+from hazelcast.invocation import Invocation
+from hazelcast.util import (
+    UUIDUtil,
+    check_not_none,
+    to_millis,
+    check_true,
+    get_attr_name,
+    try_to_get_error_message,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class SqlService(object):
+    """A service to execute SQL statements.
+
+    The service allows you to query data stored in
+    :class:`Map <hazelcast.proxy.map.Map>`.
+
+    Warnings:
+
+        The service is in beta state. Behavior and API might change
+        in future releases.
+
+    **Querying an IMap**
+
+    Every Map instance is exposed as a table with the same name in the
+    ``partitioned`` schema. The ``partitioned`` schema is included into
+    a default search path, therefore a Map could be referenced in an
+    SQL statement with or without the schema name.
+
+    **Column resolution**
+
+    Every table backed by a Map has a set of columns that are resolved
+    automatically. Column resolution uses Map entries located on the
+    member that initiates the query. The engine extracts columns from a
+    key and a value and then merges them into a single column set.
+    In case the key and the value have columns with the same name, the
+    key takes precedence.
+
+    Columns are extracted from objects as follows (which happens on the
+    server-side):
+
+    - For non-Portable objects, public getters and fields are used to
+      populate the column list. For getters, the first letter is converted
+      to lower case. A getter takes precedence over a field in case of naming
+      conflict
+    - For :class:`Portable <hazelcast.serialization.api.Portable>` objects,
+      field names used in the
+      :func:`write_portable() <hazelcast.serialization.api.Portable.write_portable>`
+      method are used to populate the column list
+
+    The whole key and value objects could be accessed through a special fields
+    ``__key`` and ``this``, respectively. If key (value) object has fields,
+    then the whole key (value) field is exposed as a normal field. Otherwise the
```

Comment:
Like I said if the `__key` is a complex object, it will return it as a whole. So, you cannot access them like `age`, `name` but have to use `__key.age` and `__key.name`

Reply:
My problem is with `key (value)` part. It makes the sentence complex. I would remove `(value)`

Reply:
will change it to `key (or value)`